src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Windows.cs

```diff
@@ -294,14 +294,14 @@ private unsafe int ReadCore(Span<byte> buffer)
 
         private ValueTask<int> ReadAsyncCore(Memory<byte> buffer, CancellationToken cancellationToken)
         {
-            var completionSource = new ReadWriteCompletionSource(this, buffer, isWrite: false);
+            var valueTaskSource = new ReadWriteValueTaskSource(this, buffer, isWrite: false);
```

Comment:
Part of the work for the associated issue is having a reusable instance of this cached (two actually, one for reading, one for writing) on the PipeStream instance so that most operations end up being ammortized allocation-free.  See FileStream for how it's rented.

Reply:
Done.
@adamsitnik Do we have some benchmarks available to evaluate the improvements around allocation with this PR?

Reply:
> Do we have some benchmarks available to evaluate the improvements around allocation with this PR?

We do. You can find them in the [dotnet/performance](https://github.com/dotnet/performance/tree/main/src/benchmarks/micro/libraries/System.IO.Pipes) repo. 

You can use [CoreRun](https://github.com/dotnet/performance/tree/main/src/benchmarks/micro#private-runtime-builds) to see how your changes are affecting the performance.

Reply:
|                                         Type |    Method |        Job |           Toolchain |    size |       Mean |    Error |   StdDev |     Median |        Min |        Max | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------------------------- |---------- |----------- |-------------------- |-------- |-----------:|---------:|---------:|-----------:|-----------:|-----------:|------:|--------:|------:|------:|------:|----------:|
|  Perf_AnonymousPipeStream_ServerIn_ClientOut | ReadWrite | Job-DJACCE |  \after\CoreRun.exe | 1000000 |   639.9 μs | 19.89 μs | 22.10 μs |   642.4 μs |   593.9 μs |   688.4 μs |  1.02 |    0.06 |     - |     - |     - |     139 B |
|  Perf_AnonymousPipeStream_ServerIn_ClientOut | ReadWrite | Job-WNSLDO | \before\CoreRun.exe | 1000000 |   631.2 μs | 19.08 μs | 21.20 μs |   625.3 μs |   597.6 μs |   665.8 μs |  1.00 |    0.00 |     - |     - |     - |     132 B |
|                                              |           |            |                     |         |            |          |          |            |            |            |       |         |       |       |       |           |
|  Perf_AnonymousPipeStream_ServerOut_ClientIn | ReadWrite | Job-DJACCE |  \after\CoreRun.exe | 1000000 |   643.8 μs | 12.33 μs | 13.19 μs |   642.7 μs |   609.2 μs |   671.8 μs |  1.01 |    0.04 |     - |     - |     - |     134 B |
|  Perf_AnonymousPipeStream_ServerOut_ClientIn | ReadWrite | Job-WNSLDO | \before\CoreRun.exe | 1000000 |   636.5 μs | 17.81 μs | 19.06 μs |   636.4 μs |   598.6 μs |   674.0 μs |  1.00 |    0.00 |     - |     - |     - |     132 B |
|                                              |           |            |                     |         |            |          |          |            |            |            |       |         |       |       |       |           |
| Perf_NamedPipeStream_ServerInOut_ClientInOut | ReadWrite | Job-DJACCE |  \after\CoreRun.exe | 1000000 | 1,078.6 μs | 21.88 μs | 24.32 μs | 1,073.9 μs | 1,039.4 μs | 1,118.8 μs |  1.27 |    0.05 |     - |     - |     - |     515 B |
| Perf_NamedPipeStream_ServerInOut_ClientInOut | ReadWrite | Job-WNSLDO | \before\CoreRun.exe | 1000000 |   850.4 μs | 21.47 μs | 24.72 μs |   857.1 μs |   801.7 μs |   879.8 μs |  1.00 |    0.00 |     - |     - |     - |     926 B |
|                                              |           |            |                     |         |            |          |          |            |            |            |       |         |       |       |       |           |
|      Perf_NamedPipeStream_ServerIn_ClientOut | ReadWrite | Job-DJACCE |  \after\CoreRun.exe | 1000000 | 1,184.5 μs | 82.79 μs | 95.35 μs | 1,173.7 μs | 1,031.6 μs | 1,327.6 μs |  1.45 |    0.11 |     - |     - |     - |     527 B |
|      Perf_NamedPipeStream_ServerIn_ClientOut | ReadWrite | Job-WNSLDO | \before\CoreRun.exe | 1000000 |   823.5 μs | 16.43 μs | 18.26 μs |   821.0 μs |   783.6 μs |   860.9 μs |  1.00 |    0.00 |     - |     - |     - |     919 B |
|                                              |           |            |                     |         |            |          |          |            |            |            |       |         |       |       |       |           |
|      Perf_NamedPipeStream_ServerOut_ClientIn | ReadWrite | Job-DJACCE |  \after\CoreRun.exe | 1000000 | 1,051.3 μs | 32.30 μs | 37.20 μs | 1,052.2 μs |   981.8 μs | 1,116.3 μs |  1.27 |    0.04 |     - |     - |     - |     517 B |
|      Perf_NamedPipeStream_ServerOut_ClientIn | ReadWrite | Job-WNSLDO | \before\CoreRun.exe | 1000000 |   831.4 μs | 15.77 μs | 14.75 μs |   827.5 μs |   808.8 μs |   858.1 μs |  1.00 |    0.00 |     - |     - |     - |     919 B |

Reply:
I am not sure how to interpret these results...
- There is nearly no impact on `AnonymousPipeStream`.
- We observe a huge allocation drop for `NamedPipeStream` but at the price of a huge duration bump !?!

Can you please help me to understand what happens here?

Reply:
> There is nearly no impact on AnonymousPipeStream.

Anonymous pipes don't support overlapped I/O, so the code path that uses this isn't ever hit there.

> We observe a huge allocation drop for NamedPipeStream but at the price of a huge duration bump !?!

I'm skeptical of those numbers... I expect there's something amiss in either the implementation or benchmark.  I still need to review the PR.